File: backend-spring/src/main/java/com/nearkart/entity/SalePayment.java
package com.nearkart.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "sale_payments")
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor
@Builder
public class SalePayment {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Column(name = "company_id")
    private UUID companyId;

    @Column(name = "buyer_id", nullable = false)
    private UUID buyerId;

    @Column(name = "payment_date")
    private LocalDate paymentDate;

    @Column(name = "invoice_no")
    private String invoiceNo;

    @Column(nullable = false, precision = 14, scale = 2)
    private BigDecimal total;

    @Column(nullable = false, precision = 14, scale = 2)
    private BigDecimal received;

    @Column(nullable = false, precision = 14, scale = 2)
    private BigDecimal balance;

    @Column(name = "bank_name")
    private String bankName;

    @Column(name = "cheque_no")
    private String chequeNo;

    @Column(name = "payment_mode")
    @Builder.Default
    private String paymentMode = "cash";

    @Column(columnDefinition = "TEXT")
    private String narration;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "buyer_id", insertable = false, updatable = false)
    private Buyer buyer;

    @PrePersist
    protected void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        createdAt = now; updatedAt = now;
        if (paymentDate == null) paymentDate = LocalDate.now();
    }

    @PreUpdate
    protected void onUpdate() { updatedAt = OffsetDateTime.now(); }
}

File: backend/app/schemas/sale_payment.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class SalePaymentCreate(BaseModel):
    company_id: UUID | None = None
    buyer_id: UUID
    payment_date: date
    invoice_no: str | None = None
    total: Decimal
    received: Decimal
    balance: Decimal
    bank_name: str | None = None
    cheque_no: str | None = None
    payment_mode: str = "cash"
    narration: str | None = None


class SalePaymentUpdate(BaseModel):
    company_id: UUID | None = None
    buyer_id: UUID | None = None
    payment_date: date | None = None
    invoice_no: str | None = None
    total: Decimal | None = None
    received: Decimal | None = None
    balance: Decimal | None = None
    bank_name: str | None = None
    cheque_no: str | None = None
    payment_mode: str | None = None
    narration: str | None = None


class SalePaymentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    company_id: UUID | None
    buyer_id: UUID
    payment_date: date
    invoice_no: str | None
    total: Decimal
    received: Decimal
    balance: Decimal
    bank_name: str | None
    cheque_no: str | None
    payment_mode: str
    narration: str | None
    created_at: datetime
    updated_at: datetime
